### PR TITLE
Push variables to console when instantiated

### LINKED ISSUE #4098: `update_console` example is broken
## 🐛 Bug

The `update_console` example: https://github.com/napari/napari/blob/main/examples/update_console.py does not acutally filly the ipython console with variables.

Also discussed here: https://napari.zulipchat.com/#narrow/stream/212875-general/topic/ipython-variables-from-script

## To Reproduce

1. Run the example
2. try to use some of the locals in the napari ipython console

## Expected behavior
Locals are there

## Environment

```
napari: 0.4.14
Platform: Linux-5.15.15-76051515-generic-x86_64-with-glibc2.34
System: Pop!_OS 21.10
Python: 3.9.10 | packaged by conda-forge | (main, Jan 30 2022, 18:04:04) [GCC 9.4.0]
Qt: 5.12.9
PyQt5: 5.12.3
NumPy: 1.22.1
SciPy: 1.7.3
Dask: 2022.01.1
VisPy: 0.9.4

OpenGL:
- GL version: 4.6 (Compatibility Profile) Mesa 21.2.2
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1080, scale 1.0

Plugins:
- console: 0.0.4
- scikit-image: 0.4.14
- svg: 0.1.5
```
## Additional context

I believe that this is due to the lazy loading of the console. In particular here:

https://github.com/napari/napari/blob/bf8682ec9e1ba50e40f3d078a68360903ce58a06/napari/viewer.py#L78-L81

the check for `_console` will always return `None` here as the console hasn't been initialized. You can forcibly init the console by accessing the

`_qt_viewer.console` property, but that's not ideal because of the usage of a private variable. So i suggest changing the beahvior of `update_console` to:

1. `update_console` defaults to intiatlizing the console if one doesn't exist
2.  add a kwarg to brign the current behavior of not creating one

I think this is a sensible default because if you are updating the console then you probably want one to exist if it doesn't already.

### PR DESCRIPTION
**Pull request by davidpross**
# Description

Currently, if [`update_console`](https://github.com/napari/napari/blob/7117d22e4baa740864762454911dc5ed09764ccc/napari/viewer.py#L88) is called on viewer object before the console is instantiated the function does nothing. With this PR instead references to the requested variables are added to a `console_backlog` list attribute on qt_viewer. If and when console is instantiated these variables are pushed.

Unsure if [this related PR](https://github.com/napari/napari-console/pull/18) is moving forward and whether it would solve this case.

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)

# References

Closes #4098

# How has this been tested?

- [x] Ran unit tests for `napari/_qt`. There were some errors but looked unrelated to my changes.
- [x] Added unit test based on existing `test_update_console`

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by psobolewskiPhD** [comment]
This works quite nicely for me locally!
While you have this PR open, can you also fix the example that the issue refers to? 
It's making two viewers, in a rather nonsensical fashion, as noted in the original issue:
https://github.com/napari/napari/issues/4098#issuecomment-1041919544
Worse yet, it means using the example to test your fix doesn't work!

Just change:
https://github.com/napari/napari/blob/c305b4607569b963ca4862172e9ef8a9715b52bc/examples/update_console.py#L22
from `napari.view_image` to `viewer.add_image`
(Also, I'm not sure what the deal is with printing in the doc string, it doesn't print anything, but with your fix one can access `polygons`. You could delete that I guess.)

I think fixing this is useful, even if I personally think the console should inherit from scripts automagically. That's what https://github.com/napari/napari-console/pull/18 is trying to do, which would make this obsolete, but it's still WIP, so in the meantime this is nice.

**Comment by davidpross** [comment]
> This works quite nicely for me locally!
> While you have this PR open, can you also fix the example that the issue refers to?

Thanks for checking it out, I added a fix for the example to not open two viewers and also printed the updated shapes coordinates after the window closes.

**Comment by Czaki** [comment]
Aaybe it will be also nice to always push `np` to the console to not require type `import numpy as np`

**Comment by psobolewskiPhD** [comment]
> Aaybe it will be also nice to always push `np` to the console to not require type `import numpy as np`

On one hand I agree, it's handy, saves keystrokes. On the other it seems potentially dangerous because some snippets will work as if by magic without an explicit import and a new python user may then be confused when in ipython/jupyter they don't. I've found the console very nice to learn and explore, as well as show python to people, so I think i'd prefer that such behavior not be included.
Perhaps it's a new PR to make some sort of console init preference to import on open?

**Comment by brisvag** [comment]
> On the other it seems potentially dangerous because some snippets will work as if by magic without an explicit import and a new python user may then be confused when in ipython/jupyter they don't.

In the past I dealt with similar issue by having the console print a header where it informs that some things have been imported for the user, something like:

```
--- napari console ---
The napari viewer is accessible from the `viewer` variable.
Additionally, numpy was imported as `np` and ...
---
```

**Comment by psobolewskiPhD** [comment]
I like that a lot. Maybe 2nd PR though? Could be some discussion as to what should be auto-imported and how the message should look.

**Comment by hanjinliu** [comment]
I think the safer but more useful way is to execute the ipython startup files on launching the napari console (see [here](https://switowski.com/blog/ipython-startup-files/) as an example if you are not familiar with).

**Comment by brisvag** [comment]
Yes, that's also great, though not really equivalent (we still have the `viewer` variable, for example, which is a napari-only thing)

**Comment by jni** [comment]
I think napari/napari-console#18 would indeed solve this. I'd be in favour of pushing on that one but I'm not sure whether @Carreau is around to push it over the line... I'm not opposed to merging this though because the implementation is so simple. I only have a minor question which is that it seems a bit pointless to reset the blank value to `None` instead of always using the empty list? ie you would save an if clause if you just set it to the empty list?

File: napari/_tests/test_advanced.py [hunk 1]
```diff
@@ -180,6 +180,34 @@ def test_update_console(make_napari_viewer):
         del viewer.window._qt_viewer.console.shell.user_ns[k]
 
 
+@pytest.mark.filterwarnings("ignore::DeprecationWarning:jupyter_client")
+def test_update_lazy_console(make_napari_viewer):
```
Comment (kephale) [inline napari/_tests/test_advanced.py]: I am trying to understand if this test will still do its job if the startup time for console is decreased to the point that it is loaded when `update_console` is called.

Comment (davidpross) [reply napari/_tests/test_advanced.py]: With the [lazy instantiation](https://github.com/napari/napari/blob/b5696770d31cbaa6dc02068b8b8b0a95bdd8a107/napari/_qt/qt_viewer.py#L489) the console shouldn't begin starting up until the assertion on L196 below.

**Comment by davidpross** [comment]
> I only have a minor question which is that it seems a bit pointless to reset the blank value to `None` instead of always using the empty list? ie you would save an if clause if you just set it to the empty list?

I think my [last commit](https://github.com/napari/napari/pull/5210/files/16d362a382964587ee7552056dff215478efd406..fd50eb9b632be4e6d71095d35ca39ea20f0ecff1) implemented your suggestion but let me know if I misinterpreted it.

**Comment by Carreau** [comment]
> I think [napari/napari-console#18](https://github.com/napari/napari-console/pull/18) would indeed solve this. I'd be in favor of pushing on that one but I'm not sure whether @Carreau is around to push it over the line...

Yeah, sorry I'm been a bit under the wether over the past two weeks. 

Seeing that napari console is not super active, have considered having a mono-repo under napari, that has both napari and napari console, it would make it a bit easier to manage all in the same place ?

**Comment by Carreau** [comment]
I've updated https://github.com/napari/napari-console/pull/18, it works as well, but i'm not again this also getting in, as https://github.com/napari/napari-console/pull/18 will only capture variable in in the direct calling scope, and I can see how someone would like to push variables from a different place.

**Comment by psobolewskiPhD** [comment]
> Seeing that napari console is not super active, have considered having a mono-repo under napari, that has both napari and napari console, it would make it a bit easier to manage all in the same place ?

I raised this on zulip, perhaps in the dev stream, I forgot. The consensus was to devolve more of napari to separate repos rather than rolling things back in. 
@jni can correct me, but my understanding is that the platonic ideal is a lean napari-core and everything else being optional.

**Comment by Carreau** [comment]
> @jni can correct me, but my understanding is that the platonic ideal is a lean napari-core and everything else being optional.

Sorry if I was unclear, I agree that napari and napari-console should be different packages, I advocate for many packages per repos, kind-of the way jupyter-lab does it.

**Comment by davidpross** [comment]
@Carreau @jni After some further reflection I'm not sure whether [napari/napari-console#18](https://github.com/napari/napari-console/pull/18) will solve the issue that led me to create this PR. For my napari plugin when you drag a folder of files or a file in that the plugin is registered to read the plugin creates a model class object and also auto adds a dock widget to the napari viewer. The plugin also wants to add a reference to this model class object to the console because you can inspect things there and call custom functions that are not accessible from the dock widget. But if I call `update_console` and the console isn't instantiated that reference isn't added when the console is later opened. I've solved it in my case by triggering the `update_console` call again when the dockConsole.visibilityChanged event occurs but I thought the PR here might be a cleaner approach. I haven't tested the change to `napari-console` but I don't think that would cover this scenario.

File: napari/layers/surface/_tests/test_surface.py [hunk 2]
```diff
@@ -311,7 +311,7 @@ def test_surface_normals():
     faces = np.array([[0, 1, 2], [3, 4, 5], [6, 7, 8]])
     values = np.array([1, 2, 3, 1, 2, 3, 1, 2, 3])
 
-    normals = dict(face=dict(visible=True, color='red'))
+    normals = {"face": {"visible": True, "color": 'red'}}
```
Comment (jni) [inline napari/layers/surface/_tests/test_surface.py]: Can you revert these unrelated changes @davidpross?

File: napari/_qt/qt_viewer.py [hunk 3]
```diff
@@ -497,6 +503,9 @@ def console(self):
                     self.console.push(
                         {'napari': napari, 'action_manager': action_manager}
                     )
+                    for i in self.console_backlog:
+                        self.console.push(i)
+                    self._console_backlog = []
```
Comment (jni) [inline napari/_qt/qt_viewer.py]: @Czaki and I discussed this idea at the euro/asia community meeting yesterday. @Czaki's concern (if I'm not butchering it) is that this list could lead to memory "leaks" if a plugin adds things and then the plugin is closed down without the console ever getting called. His suggestion (again I might be butchering it) is to use *weakrefs* when adding to the backlog, and then here you convert them to real refs. Does that make sense?

**Review by jni — CHANGES_REQUESTED** [review]
@davidpross interesting point! I agree that napari/napari-console#18 has nothing to do with your use case it turns out! 😅 

Please see below for some concerns — what do you think about the weakref option?

File: napari/layers/surface/_tests/test_surface.py (hunk 2 above)
Comment (psobolewskiPhD) [reply napari/layers/surface/_tests/test_surface.py]: Those are from the new ruff linter I think?

Comment (jni) [reply napari/layers/surface/_tests/test_surface.py]: Yes I recognise them, but why are they in this PR? Nothing else touches that file?

Comment (psobolewskiPhD) [reply napari/layers/surface/_tests/test_surface.py]: Good question. I merged main to have tests pass and pre-commit fired.
Maybe a rule change in the linter? maybe @Czaki has an idea?

Comment (Czaki) [reply napari/layers/surface/_tests/test_surface.py]: I have merged the main branch as this change was already there.

File: napari/_qt/qt_viewer.py (hunk 3 above)
Comment (davidpross) [reply napari/_qt/qt_viewer.py]: Yes, makes sense, I was wondering about memory issues, too, but didn't know a solution. I'll look into this option.

File: napari/layers/surface/_tests/test_surface.py (hunk 2 above)
Comment (davidpross) [reply napari/layers/surface/_tests/test_surface.py]: Thanks for figuring it out, I didn't recognize this change.

File: napari/_qt/qt_viewer.py (hunk 3 above)
Comment (davidpross) [reply napari/_qt/qt_viewer.py]: I added in logic to make weakrefs when possible when adding to backlog. Let me know if things look OK.

**Review by jni — APPROVED** [review]
Thanks! I don't know under what conditions `ref` fails, so that would be good to add to the `_weak_referenceable` docstring.

I also would change the following in the code:

`_weak_refereanceable` ->

```python
def _weakref_if_possible(obj):
    try:
        newref = ref(obj)
    except TypeError:  # unable to make weak reference because...
        newref = obj
    return newref
```

And instead of new dicts:

```python
def _unwrap_if_weakref(value):
    if isinstance(value, ref):
        return value()
    else:
        return value
```

Then you can get rid of a lot of ifs: add `_weakref_if_possible(obj)` always to the backlog, and add `_unwrap_if_weakref(value)` to the console.

@Czaki what do you think of this new implementation?

**Comment by davidpross** [comment]
Added description for when creating a weakref fails, also made suggested updates. In addition on further reflection removed special handling for list objects because if a list or tuple is passed it is supposed to only contain strings of variable names. So weakrefs are created if possible for the values of any dictionary passed to `update_console` when console hasn't been instantiated. A weakref will be created for something like a numpy array but not a Python list. So there could still be some strong references if you passed something like `update_console({'obj_list': [obj1, obj2, obj3]})`

File: napari/_qt/qt_viewer.py [hunk 4]
```diff
@@ -482,6 +483,70 @@ def _create_performance_dock_widget(self):
...
+        return newref
+
+    def _unwrap_if_weakref(self, value):
+        """Returns value or if that is weakref the object referenced by value.
+
+        Args:
+            value (Object|weakref): Can be a weakref
+
+        Returns:
+            Object|None: Returns object or None if weakref is dead.
```
Comment (jni) [inline napari/_qt/qt_viewer.py]: @davidpross can you use numpydoc formatting here? 🙏 (and above). Otherwise, nice, thanks! 😊

Comment (davidpross) [reply napari/_qt/qt_viewer.py]: Updated docstrings, thanks for all the help with these changes.

**Comment by kephale** [comment]
@Carreau just a gently nudge for review on this one. Thank you!

**Comment by Carreau** [comment]
> @Carreau just a gently nudge for review on this one. Thank you!

That looks good to me, there will always be edge case we can refine later.

**Review by jni — APPROVED** [review]

File: napari/_qt/qt_viewer.py [hunk 5]
```diff
@@ -452,6 +527,19 @@ def console(self):
...
+                            # recover weak refs
+                            self.console.push(
+                                {
+                                    k: self._unwrap_if_weakref(v)
+                                    for k, v in i.items()
+                                    if self._unwrap_if_weakref(v) is not None
+                                }
+                            )
+                        else:
+                            self.console.push(i)
```
Comment (jni) [inline napari/_qt/qt_viewer.py]: @davidpross I think this line isn't covered by tests? Ditto for line 508. tbh I don't understand the usage that would get us there. Could you elaborate on that and maybe add a test? I think this is very close to merging otherwise! 🙏

**Comment by jni** [comment]
btw I've pushed some minor formatting fixes so please `git pull` before making further changes!

File: napari/_qt/qt_viewer.py (hunk 5 above)
Comment (davidpross) [reply napari/_qt/qt_viewer.py]: Good catch on the test coverage. It turns out that passing a variable name as a string or list of variable names was not working in the context of a console that is not yet instantiated. I added code from [here](https://github.com/ipython/ipython/blob/b3cd3b056a90b5d553746e96628aeb6dbb7b37dd/IPython/core/interactiveshell.py#L1517) to handle these cases and updated the unit tests accordingly. Also the console_backlog will now always be a list of dicts.
